components/debounce/browser/debounce_download_service.cc

```diff
@@ -0,0 +1,255 @@
+
+  return true;
+}
+
+}  // namespace
+
+DebounceRule::DebounceRule() = default;
+
+DebounceRule::~DebounceRule() = default;
+
+void DebounceRule::clear() {
+  include_pattern_set_.ClearPatterns();
+  exclude_pattern_set_.ClearPatterns();
+  action_ = kDebounceNoAction;
+  param_ = "";
+}
+
+void DebounceRule::Parse(base::ListValue* include_value,
+                         base::ListValue* exclude_value,
+                         const std::string& action,
+                         const std::string& param) {
+  clear();
+  std::string error;
+
+  // Debouncing only affects HTTP or HTTPS URLs, regardless of how the rules are
+  // written. (Also, don't write rules for other URL schemes, because they won't
+  // work and you're just wasting everyone's time.)
+  int valid_schemes = URLPattern::SCHEME_HTTP | URLPattern::SCHEME_HTTPS;
+  if (!include_pattern_set_.Populate(*include_value, valid_schemes, false,
+                                     &error) ||
+      !exclude_pattern_set_.Populate(*exclude_value, valid_schemes, false,
+                                     &error)) {
+    LOG(ERROR) << error;
+    clear();
+    return;
+  }
+  if (action == "redirect")
+    action_ = kDebounceRedirectToParam;
+  else if (action == "base64,redirect")
+    action_ = kDebounceBase64DecodeAndRedirectToParam;
+  else if (action == "remove")
+    action_ = kDebounceRemoveParam;
+  param_ = param;
+}
+
+bool DebounceRule::Apply(const GURL& original_url,
+                         const net::SiteForCookies& original_site_for_cookies,
+                         GURL* final_url) {
+  // If URL matches an explicitly excluded pattern, this rule does not apply.
+  if (exclude_pattern_set_.MatchesURL(original_url))
+    return false;
+
+  // If URL does not match an explicitly included pattern, this rule does not
+  // apply.
+  if (!include_pattern_set_.MatchesURL(original_url))
+    return false;
+
+  if (action_ == kDebounceRemoveParam) {
+    // RemoveQueryParameter is guaranteed to return a valid URL in
+    // |final_url|, so no need for an is_valid check here. It also
+    // returns false if the given parameter was not found, which we
+    // pass to the caller to say that the rule was not applied.
+    return RemoveQueryParameter(original_url, param_, final_url);
+  }
+
+  if (action_ == kDebounceRedirectToParam ||
+      action_ == kDebounceBase64DecodeAndRedirectToParam) {
+    std::string unescaped_value;
+    if (!net::GetValueForKeyInQuery(original_url, param_, &unescaped_value))
+      return false;
+    GURL new_url;
+    if (action_ == kDebounceBase64DecodeAndRedirectToParam) {
+      std::string base64_decoded_value;
+      if (!base::Base64Decode(unescaped_value, &base64_decoded_value))
+        return false;
+      new_url = GURL(base64_decoded_value);
+    } else {
+      new_url = GURL(unescaped_value);
+    }
+
+    // Failsafe: ensure we got a valid URL out of the param.
+    if (!new_url.is_valid())
```

Comment:
In addition to making sure that the extract URL is valid, we should only use it if its scheme is HTTP or HTTPS.